### PR TITLE
nixos/manual: translate manpages to mdoc

### PR DESCRIPTION
**Pull request by pennae**
###### Description of changes

since we want to move away from docbook and having docbook manpages around is going to block further progress we have to translate the current nixos manpages from docbook it *something* else. mdoc seems the most appropriate choice since markdown can't represent all the things manpages can differentiate with even more extensions. if we ever need html manpages we can render them using troff, like many of the online manpage viewers, or rewrite them again. since we haven't had html manpages for any of these in many years that seems unlikely to happen.

unlike most of the recent markdown conversions this comes with a lot of minor rendering changes, mostly in spacing. docbook-xslt creates manual pages in a very different dialect than mdoc (which is used here).

also tagging @alyssais and @sternenseemann since both suggested using mdoc.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/man-nixos-build-vms.8 [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+.Os NixOS
+.Sh NAME
+.Nm nixos-build-vms
+.Nd build a network of virtual machines from a network of NixOS configurations
+.
+.
+.
+.Sh SYNOPSIS
+.Nm nixos-build-vms
+.Op Fl -show-trace
```
Comment (alyssais) [inline nixos/doc/manual/man-nixos-build-vms.8:19]: Every time you have "Fl -" like this, you can instead write "Fl Fl", which maybe looks a bit strange. (But I don't mind if you like it better this way.)

File: nixos/doc/manual/man-nixos-option.8 [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+.
+.Pp
+When the option name is not an option, the command prints the list of attributes
+contained in the attribute set.
+.
+.
+.
+.Sh OPTIONS
+.Bl -tag -width indent
+.It Fl r | -recursive
```
Comment (alyssais) [inline nixos/doc/manual/man-nixos-option.8:35]: The current man pages use a comma between aliases in the options section, which I think looks nicer.  The space between the preceding flag and the comma will be deleted.

**Review by alyssais — COMMENTED** [review]
This is excellent work.  These are some very lovely looking man pages.

I've made a couple of comments, but both are very small details that are ultimately down to personal preference, so only change them if you agree. :)

**Comment by ncfavier** [comment]
I'm lukewarm on this. Going from DocBook to mdoc seems like a step *backwards* if the goal is to make documentation easier to write (which I think it is, rather than just to get rid of DocBook).

> since we haven't had html manpages for any of these in many years that seems unlikely to happen.

That would be nice actually. I wanted to link to a nixos-version flag the other day and couldn't.

**Comment by ncfavier** [comment]
Why not just stick to DocBook for man pages? Clearly generating nice HTML pages is possible since [this](https://www.mankier.com/8/nixos-version) exists, but I don't know how they do it.

**Review by alyssais — APPROVED** [review]

**Comment by alyssais** [comment]
I definitely don't think it's easier to produce man pages with DocBook.  For one thing, you have to figure out how to build the DocBook man pages, but with mdoc, there's no build step necessary — you just ran `man` on the file you're working with.  In either case, you likely don't need to go out of your way to learn the syntax and can copy what you need from elsewhere in the man pages.

The middle of the "good at generating man pages" and "easy for humans to write" diagram in AsciiDoc, but it would be a bit strange to have _just_ the man pages in AsciiDoc when it's so similar to markdown.  It really drives home that AsciiDoc would have been a much better choice than Markdown overall, but oh well…

File: nixos/doc/manual/default.nix [hunk 3]
```diff
@@ -256,7 +257,13 @@ in rec {
...
     ''
       # Generate manpages.
-      mkdir -p $out/share/man
+      mkdir -p $out/share/man/man8
+      (
+        cd ${mdocSource}
+        for mp in man-*.8; do
+          cp "$mp" "$out/share/man/man8/''${mp#man-}"
+        done
+      )
```
Comment (SuperSandro2000) [inline nixos/doc/manual/default.nix:266]: If the man pages are already named correct, we can just use installManPage.

**Review by SuperSandro2000 — APPROVED** [review]

File: nixos/doc/manual/manpages/nixos-rebuild.8 [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+.
+.It Fl -install-bootloader
+Causes the boot loader to be (re)installed on the device specified by the
+relevant configuration options.
+.
+.It Fl -no-build-nix
+Normally,
+.Nm
+first builds the
+.Ql nixUnstable
```
Comment (sternenseemann) [inline nixos/doc/manual/manpages/nixos-rebuild.8:232]: Is that still accurate?

File: nixos/doc/manual/manpages/nixos-rebuild.8 [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+.Pa /nix/var/nix/profiles/system Ns
+\&.)
+.
+.It Fl -builders Ar builder-spec
+Allow ad-hoc remote builders for building the new system. This requires
+the user executing
+.Nm
+(usually root) to be configured as a trusted user in the Nix daemon. This can be
+achieved by using the
+.Ql nix.settings.trusted-users
```
Comment (sternenseemann) [inline nixos/doc/manual/manpages/nixos-rebuild.8:259]:
```suggestion
.Va nix.settings.trusted-users
```

File: nixos/doc/manual/manpages/nixos-rebuild.8 [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+.Dq NixOS - Profile 'test' Ns
+\&.
+.
+.It Fl -specialisation Ar name , Fl c Ar name
+Activates given specialisation; when not specified, switching and testing
+will activate the base, unspecialised system.
+.
+.It Fl -build-host Ar host
+Instead of building the new configuration locally, use the specified host
+to perform the build. The host needs to be accessible with ssh, and must
```
Comment (sternenseemann) [inline nixos/doc/manual/manpages/nixos-rebuild.8:297]:
```suggestion
to perform the build. The host needs to be accessible with
.Ic ssh Ns ,
and must
```

File: nixos/doc/manual/manpages/nixos-rebuild.8 [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+You can include a remote user name in the host name
+.Ns ( Va user@host Ns
+). You can also set ssh options by defining the
+.Ev NIX_SSHOPTS
+environment variable.
+.Pp
+Note that
+.Nm
+honors the
+.Ql nixpkgs.crossSystem
```
Comment (sternenseemann) [inline nixos/doc/manual/manpages/nixos-rebuild.8:339]:
```suggestion
.Va nixpkgs.crossSystem
```

File: nixos/doc/manual/manpages/nixos-rebuild.8 [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+Allow ad-hoc remote builders for building the new system. This requires
+the user executing
+.Nm
+(usually root) to be configured as a trusted user in the Nix daemon. This can be
+achieved by using the
+.Ql nix.settings.trusted-users
+NixOS option. Examples values for that option are described in the
+.Dq Remote builds
+chapter in the Nix manual, (i.e.
+.Ql --builders \(dqssh://bigbrother x86_64-linux\(dq Ns
```
Comment (sternenseemann) [inline nixos/doc/manual/manpages/nixos-rebuild.8:263]: Use `Fl` here?

File: nixos/doc/manual/manpages/nixos-rebuild.8 [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+). By specifying an empty string existing builders specified in
+.Pa /etc/nix/machines
+can be ignored:
+.Ql --builders \(dq\(dq
```
Comment (sternenseemann) [inline nixos/doc/manual/manpages/nixos-rebuild.8:267]: `Fl` here?

**Review by sternenseemann — APPROVED** [review]
Thanks for doing this! I think using mdoc is the best solution for man pages by far. Convenience improvements like splitting out common sections can still be achieved using a very simple build system.

Adding a README for documentation writers would be nice which tells them about [mdoc(7)](https://man.openbsd.org/mdoc) and when to use e.g. `Va`, `Pa`, `Ql`, … so that stuff is consistent.

The man pages could obviously still be improved a lot, but that is besides the point since this PR is just for conversion. Feel free to ignore any suggestions that are besides the point of this PR that slipped in…

**Review by pennae — COMMENTED** [review]
thank you for the extremely detailed review @sternenseemann!

File: nixos/doc/manual/manpages/nixos-generate-config.8 [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+.Sh EXAMPLES
+This command is typically used during NixOS installation to write initial
+configuration modules. For example, if you created and mounted the target file
+systems on
+.Pa /mnt
+and
+.Pa /mnt/boot Ns
+, you would run:
+.Bd -literal -offset indent
+$ </prompt>nixos-generate-config --root /mnt
```
[the remark this answers is not in the document]
Comment (sternenseemann) [reply nixos/doc/manual/manpages/nixos-generate-config.8]: Fair enough.

File: nixos/doc/manual/manpages/nixos-rebuild.8 (hunk 8 above)
Comment (sternenseemann) [reply nixos/doc/manual/manpages/nixos-rebuild.8:263]: You did use it for inline flags with example arguments in other man pages, hence the comment.

File: nixos/doc/manual/manpages/nixos-rebuild.8 [hunk 11]
```diff
@@ -0,0 +1,454 @@
+.Dd January 1, 1980
+.\" nixpkgs groff will use Nixpkgs the OS in the title by default, taking it from
+.\" doc-default-operating-system. mandoc doesn't have this register set by default,
+.\" so we can use it as a groff/mandoc switch.
+.ie ddoc-default-operating-system .Dt nixos-rebuild \&8 "NixOS System Manager's Manual"
+.el .Dt nixos-rebuild 8
+.Os NixOS
```
[the remark this answers is not in the document]
Comment (sternenseemann) [reply nixos/doc/manual/manpages/nixos-rebuild.8:7]: Shouldn't we fix it on that side and care about this later? `mandoc`'s `man` seems to be able to display it correctly, so we'd just need to fix groff/man-db.

File: nixos/doc/manual/manpages/README.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
+# NixOS manpages
+
+This is the collection of NixOS manpages, excluding `configuration.nix(5)`.
+
+Man pages are written in [`mdoc(7)` format](https://mandoc.bsd.lv/man/mdoc.7.html) and should be portable between mandoc and groff for rendering (though minor differences may occur, mandoc and groff seem to have slightly different spacing rules.)
+
+Being written in `mdoc` these manpages use semantic markup. This file is provides a guideline on where to apply which of the semantic elements of `mdoc`.
```
Comment (ncfavier) [inline nixos/doc/manual/manpages/README.md:7]:
```suggestion
Being written in `mdoc` these manpages use semantic markup. This file provides a guideline on where to apply which of the semantic elements of `mdoc`.
```

File: nixos/doc/manual/manpages/README.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+### Command lines and arguments
+
+In any manpage commands, flags and arguments to the *current* executable should be marked according to their semantics. Commands, flags and arguments passed to *other* executables should not be marked like this and should instead be considered as code examples and marked with `Ql`.
```
Comment (ncfavier) [inline nixos/doc/manual/manpages/README.md:11]:
```suggestion
In any manpage, commands, flags and arguments to the *current* executable should be marked according to their semantics. Commands, flags and arguments passed to *other* executables should not be marked like this and should instead be considered as code examples and marked with `Ql`.
```

**Review by ncfavier — COMMENTED** [review]
> added a `README.md` to the manpages folder

🙄 I still think this is unwarranted complexity but oh well.

**Pull request merged**